src/scvi/external/decipher/_trainingplan.py

```diff
@@ -0,0 +1,143 @@
+        if self.current_epoch > 0:
+            # freeze the batch norm layers after the first epoch
+            # 1) the batch norm layers helps with the initialization
+            # 2) but then, they seem to imply a strong normal prior on the latent space
+            for module in self.module.modules():
+                if isinstance(module, torch.nn.BatchNorm1d):
+                    module.eval()
+
+    def training_step(self, batch, batch_idx):
+        """Training step for Pyro training."""
+        args, kwargs = self.module._get_fn_args_from_batch(batch)
+
+        # pytorch lightning requires a Tensor object for loss
+        loss = torch.Tensor([self.svi.step(*args, **kwargs)])
+        n_obs = args[0].shape[0]
+
+        _opt = self.optimizers()
+        _opt.step()
+
+        out_dict = {"loss": loss, "n_obs": n_obs}
+        self.training_step_outputs.append(out_dict)
+        return out_dict
+
+    def on_train_epoch_end(self):
+        """Training epoch end for Pyro training."""
+        outputs = self.training_step_outputs
+        elbo = 0
+        n_obs = 0
+        for out in outputs:
+            elbo += out["loss"]
+            n_obs += out["n_obs"]
+        elbo /= n_obs
+        self.log("elbo_train", elbo, prog_bar=True)
+        self.training_step_outputs.clear()
+
+    def validation_step(self, batch, batch_idx):
+        """Validation step for Pyro training."""
+        args, kwargs = self.module._get_fn_args_from_batch(batch)
+        loss = self.differentiable_loss_fn(
+            self.scale_fn(self.module.model),
+            self.scale_fn(self.module.guide),
+            *args,
+            **kwargs,
+        )
+        nll = -self.module.predictive_log_likelihood(*args, **kwargs, n_samples=5)
+        out_dict = {"loss": loss, "nll": nll, "n_obs": args[0].shape[0]}
+        self.validation_step_outputs.append(out_dict)
+        return out_dict
+
+    def on_validation_epoch_end(self):
+        """Validation epoch end for Pyro training."""
+        outputs = self.validation_step_outputs
+        elbo = 0
+        nll = 0
+        n_obs = 0
+        for out in outputs:
+            elbo += out["loss"]
+            nll += out["nll"]
+            n_obs += out["n_obs"]
+        elbo /= n_obs
+        nll /= n_obs
+        self.log("elbo_validation", elbo, prog_bar=True)
+        self.log("nll_validation", nll, prog_bar=False)
+        self.validation_step_outputs.clear()
+
+    def configure_optimizers(self):
+        """Shim optimizer for PyTorch Lightning.
+
+        PyTorch Lightning wants to take steps on an optimizer
+        returned by this function in order to increment the global
+        step count. See PyTorch Lighinting optimizer manual loop.
+
+        Here we provide a shim optimizer that we can take steps on
+        at minimal computational cost in order to keep Lightning happy :).
+        """
+        return torch.optim.Adam([self._dummy_param])
+
+    def optimizer_step(self, *args, **kwargs):
```

Comment:
Are these missing from LowLevelPyroTrainingPlan? Should we add them there?

Reply:
these are only in the PyroTrainingPlan, not the LowLevelPyroTrainingPlan. Whether we want to push it back into the low-level class is up to you